beacon/goclient/attest.go

```diff
@@ -27,22 +27,106 @@ func (gc *GoClient) AttesterDuties(ctx context.Context, epoch phase0.Epoch, vali
+		// would be unsynchronised concurrent write (since it's cached for other concurrent readers
+		// to access). Hence, we return shallow copy here. We don't need to return deep copy because
+		// the callers of GetAttestationData will only use this data to read it (they won't update it).
+		result = &phase0.AttestationData{
+			Slot:            result.Slot,
+			Index:           committeeIndex,
+			BeaconBlockRoot: result.BeaconBlockRoot,
+			Source:          result.Source,
+			Target:          result.Target,
+		}
+	}()
+
+	// Check cache.
+	cachedResult, ok := gc.attestationDataCache.Get(slot)
+	if ok {
+		return cachedResult, spec.DataVersionPhase0, nil
+	}
+
+	// Have to make beacon node request and cache the result.
 	attDataReqStart := time.Now()
-	resp, err := gc.client.AttestationData(gc.ctx, &api.AttestationDataOpts{
-		Slot:           slot,
-		CommitteeIndex: committeeIndex,
-	})
+	result, err = func() (*phase0.AttestationData, error) {
+		// Requests with the same slot number must lock on the same mutex.
+		reqMu := &gc.attestationReqMuPool[int64(slot)%int64(len(gc.attestationReqMuPool))]
+		reqMu.Lock()
+		defer reqMu.Unlock()
+
+		// Prevent making more than 1 beacon node requests in case somebody has already made this
+		// request concurrently and succeeded.
+		cachedResult, ok := gc.attestationDataCache.Get(slot)
+		if ok {
+			return cachedResult, nil
+		}
+
+		resp, err := gc.client.AttestationData(gc.ctx, &api.AttestationDataOpts{
+			Slot: slot,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed to get attestation data: %w", err)
+		}
+		if resp == nil {
+			return nil, fmt.Errorf("attestation data response is nil")
+		}
+
+		gc.attestationDataCache.Set(slot, resp.Data)
+		gc.recentAttestationSlot.Store(uint64(slot))
+
+		return resp.Data, nil
+	}()
+	metricsAttesterDataRequest.Observe(time.Since(attDataReqStart).Seconds())
 	if err != nil {
-		return nil, DataVersionNil, fmt.Errorf("failed to get attestation data: %w", err)
-	}
-	if resp == nil {
-		return nil, DataVersionNil, fmt.Errorf("attestation data response is nil")
+		return nil, DataVersionNil, err
 	}
 
-	metricsAttesterDataRequest.Observe(time.Since(attDataReqStart).Seconds())
+	return result, spec.DataVersionPhase0, nil
+}
+
+// pruneStaleAttestationDataRunner will periodically prune attestationDataCache to keep it from growing
+// perpetually.
+func (gc *GoClient) pruneStaleAttestationDataRunner() {
+	pruneStaleAttestationData := func() {
+		// slotRetainCnt defines how many recent slots we want to preserve in attestation data cache.
+		slotRetainCnt := 5 * gc.network.SlotsPerEpoch()
+		gc.attestationDataCache.Range(func(slot phase0.Slot, data *phase0.AttestationData) bool {
+			if uint64(slot) < (gc.recentAttestationSlot.Load() - slotRetainCnt) {
+				gc.attestationDataCache.Delete(slot)
+			}
+			return true
+		})
+	}
 
-	return resp.Data, spec.DataVersionPhase0, nil
+	ticker := time.NewTicker(10 * time.Minute)
```

Comment:
Can we use 2 epoch duration here? IMO it would look nicer to clean up everything older than 2 epochs each 2 epochs

Reply:
Yes, it's easy to do, changed.